File: screen_manager/app/i18n.py
"""The languages of ESP Screens (app 0.2.90).

Every text of the app, the editor and the screens lives in one JSON file per language (screen_manager/translations,
docs/TRANSLATING.md); English is the source, and a text a language lacks shows in English. The editor reads its own
section in the browser, the screens get theirs built into their firmware (components/smart_display/screen_text_gen.py),
and this module gives the app its `addon` texts and the `screen` words it sends to the screens itself.

`Region` is Settings -> Language & region, the one place for every screen: the language (Home Assistant's, or a chosen
one), the clock (12 or 24 hours) and how numbers are written. "auto" does what Home Assistant does for a user who never
changed their profile: the clock and the numbers follow the language.
"""
import contextvars
import json
import logging
import os
import re
from pathlib import Path

LOG = logging.getLogger('screen_manager')
HERE = Path(__file__).resolve().parent
# In the app's image the files sit next to the code (Dockerfile: COPY translations); in a checkout one folder up.
FOLDER = next((path for path in (HERE / 'translations', HERE.parent / 'translations') if path.is_dir()), HERE / 'translations')
CODE = re.compile(r'[a-z]{2,3}(-[A-Za-z0-9]{2,8})?')
CLOCKS = ('auto', '24', '12')
NUMBERS = ('auto', 'point', 'comma', 'space')
PLACEHOLDER = re.compile(r'\{([a-z_]+)\}')

# Which form of a plural text ("1 hour ago | {n} hours ago") fits n; the firmware (screen_text_gen.py) and the editor
# (web/src/i18n.ts) have the same rules, docs/TRANSLATING.md says which languages use which.
def _slavic(n, one):
    t, h = n % 10, n % 100
    if one:
        return 0
    return 1 if 2 <= t <= 4 and not 12 <= h <= 14 else 2

PLURAL = {
    'one_other': lambda n: 0 if n == 1 else 1,
    'one_upto_1': lambda n: 0 if abs(n) <= 1 else 1,
    'slavic_pl': lambda n: _slavic(n, n == 1),
    'east_slavic': lambda n: _slavic(n, n % 10 == 1 and n % 100 != 11),
    'none': lambda n: 0,
}

# The language of an editor's request (its X-ESP-Screens-Language header), for the messages the app answers with.
REQUEST_LANGUAGE = contextvars.ContextVar('request_language', default='en')


def flatten(node, prefix=''):
    """(key, text) pairs in document order; a list's items get .0, .1, ... ."""
    if isinstance(node, dict):
        items = node.items()
    elif isinstance(node, list):
        items = ((str(index), value) for index, value in enumerate(node))
    else:
        return [(prefix, node)]
    pairs = []
    for key, value in items:
        pairs.extend(flatten(value, f'{prefix}.{key}' if prefix else key))
    return pairs


class Translations:
    """Every language file, read once."""

    def __init__(self, folder=FOLDER):
        self.data, self.flat = {}, {}
        for path in sorted(Path(folder).glob('*.json')):
            try:
                data = json.loads(path.read_text(encoding='utf-8'))
            except (OSError, ValueError) as error:
                LOG.warning('Translation %s is unreadable (%s); it is left out', path.name, type(error).__name__)
                continue
            if CODE.fullmatch(path.stem) and isinstance(data, dict):
                self.data[path.stem] = data
                self.flat[path.stem] = {key: value for key, value in flatten({k: v for k, v in data.items() if k != '_meta'})
                                        if isinstance(value, str)}
        self.data.setdefault('en', {'_meta': {'name': 'English', 'english': 'English', 'plural': 'one_other'}})
        self.flat.setdefault('en', {})

    def resolve(self, code):
        """The language ESP Screens has for `code`: the code itself (in any case), else its base (pt-BR -> pt), else English."""
        if not isinstance(code, str):
            return 'en'
        by_lower = {known.lower(): known for known in self.data}
        for candidate in (code, code.split('-')[0]):
            if candidate.lower() in by_lower:
                return by_lower[candidate.lower()]
        return 'en'

    def meta(self, code):
        return self.data.get(self.resolve(code), {}).get('_meta', {})

    def languages(self):
        """Every language for the settings: English first, then by English name."""
        found = [{'code': code, 'name': str(self.meta(code).get('name') or code),
                  'english': str(self.meta(code).get('english') or code), 'checked': self.meta(code).get('checked') is True}
                 for code in self.data]
        return sorted(found, key=lambda item: (not item['code'].startswith('en'), item['code'] != 'en', item['english']))

    def text(self, key, language='en', **params):
        """`key` in `language`, English where it has none, with {placeholders} filled; `n` picks a plural form."""
        language = self.resolve(language)
        text = self.flat.get(language, {}).get(key) or self.flat['en'].get(key)
        if not isinstance(text, str) or not text.strip():
            return key
        if '|' in text and 'n' in params:
            forms = [form.strip() for form in text.split('|')]
            rule = PLURAL.get(self.meta(language).get('plural'), PLURAL['one_other'])
            try:
                text = forms[min(rule(int(params['n'])), len(forms) - 1)]
            except (TypeError, ValueError):
                text = forms[-1]
        return PLACEHOLDER.sub(lambda match: str(params[match.group(1)]) if match.group(1) in params else match.group(0), text)

    def clock_24h(self, code):
        """Whether the language writes 24 hours by default (_meta.clock), as Home Assistant's "auto" does."""
        return str(self.meta(code).get('clock', '24')) != '12'

    def number_style(self, code):
        """How the language writes 1234.5: point (1,234.5), comma (1.234,5) or space (1 234,5)."""
        decimal = self.text('screen.number.decimal', code)
        group = self.text('screen.number.group', code)
        if decimal == '.':
            return 'point'
        return 'space' if group.strip() == '' else 'comma'


TRANSLATIONS = Translations()


def t(key, **params):
    """An `addon` text in the language of the editor that asked (REQUEST_LANGUAGE)."""
    return TRANSLATIONS.text(key, REQUEST_LANGUAGE.get(), **params)


# The screens' language and number format (Settings -> Language & region), for the words the app sends to the screens
# itself: tile names, the top bar, the history card. The manager keeps it current (Manager.language_changed).
SCREENS = {'language': 'en', 'numbers': 'point'}


def set_screens(language, numbers):
    SCREENS.update(language=language, numbers=numbers)


def screen_t(key, **params):
    """A text in the screens' language, for what the app sends to them."""
    return TRANSLATIONS.text(key, SCREENS['language'], **params)


def screen_number(text):
    """A number as Home Assistant sends it, written as the screens write numbers ("1.234,5" in Dutch)."""
    return format_number(text, SCREENS['numbers'])


class Region:
    """Settings -> Language & region, kept in /data/language.json: one language, clock and number format for every
    screen. A screen's firmware carries its language; the clock and the numbers travel in the layout message."""

    def __init__(self, path, translations=TRANSLATIONS, ha_language=None):
        self.path, self.tr = Path(path), translations
        self.setting, self.clock, self.numbers = 'auto', 'auto', 'auto'
        # Home Assistant's own language (get_config), once connected.
        self.ha_language = ha_language or (lambda: None)
        self.stored = self.path.exists()
        if self.stored:
            try:
                raw = json.loads(self.path.read_text(encoding='utf-8'))
            except (OSError, ValueError):
                raw = {}
            if raw.get('version') == 1:
                self.setting = raw.get('language') if raw.get('language') == 'auto' or raw.get('language') in self.tr.data else 'auto'
                self.clock = raw.get('clock') if raw.get('clock') in CLOCKS else 'auto'
                self.numbers = raw.get('numbers') if raw.get('numbers') in NUMBERS else 'auto'

    def save(self):
        self.path.parent.mkdir(parents=True, exist_ok=True)
        temporary = self.path.with_suffix('.tmp')
        with open(temporary, 'w', encoding='utf-8') as handle:
            os.chmod(temporary, 0o600)
            json.dump({'version': 1, 'language': self.setting, 'clock': self.clock, 'numbers': self.numbers}, handle)
            handle.flush()
            os.fsync(handle.fileno())
        temporary.replace(self.path)
        self.stored = True

    def change(self, data):
        """Apply {language?, clock?, numbers?} from the editor; ValueError (translated) for anything unknown."""
        if not isinstance(data, dict):
            raise ValueError(t('addon.errors.language_choice'))
        setting = data.get('setting', self.setting)
        clock = data.get('clock', self.clock)
        numbers = data.get('numbers', self.numbers)
        if setting != 'auto' and setting not in self.tr.data:
            raise ValueError(t('addon.errors.language_choice'))
        if clock not in CLOCKS or numbers not in NUMBERS:
            raise ValueError(t('addon.errors.region_choice'))
        self.setting, self.clock, self.numbers = setting, clock, numbers
        self.save()

    def language(self):
        """The language of every screen: the chosen one, else Home Assistant's, else English."""
        return self.setting if self.setting != 'auto' else self.tr.resolve(self.ha_language() or 'en')

    def clock_24h(self):
        return self.tr.clock_24h(self.language()) if self.clock == 'auto' else self.clock == '24'

    def number_style(self):
        return self.tr.number_style(self.language()) if self.numbers == 'auto' else self.numbers

    def view(self):
        """What the editor's Language & region card shows."""
        return {'setting': self.setting, 'effective': self.language(),
                'ha': self.ha_language(), 'languages': self.tr.languages(),
                'clock': self.clock, 'clock_effective': '24' if self.clock_24h() else '12',
                'numbers': self.numbers, 'numbers_effective': self.number_style()}

    def screen_text(self, key, **params):
        """A `screen` or `addon` text in the screens' language, for what the app itself sends to them."""
        return self.tr.text(key, self.language(), **params)


def format_number(text, style):
    """A number as Home Assistant sends it ("1234.5") written in `style` ("1.234,5" for comma); anything else unchanged."""
    if not isinstance(text, str) or not re.fullmatch(r'-?\d+(\.\d+)?', text):
        return text
    sign, body = ('-', text[1:]) if text.startswith('-') else ('', text)
    whole, _, rest = body.partition('.')
    group, decimal = {'point': (',', '.'), 'comma': ('.', ','), 'space': (' ', ',')}.get(style, (',', '.'))
    if len(whole) > 3:
        head = len(whole) % 3 or 3
        whole = whole[:head] + ''.join(group + whole[index:index + 3] for index in range(head, len(whole), 3))
    return sign + whole + (decimal + rest if rest else '')

File: web/src/i18n.ts
// The editor's texts (app 0.2.90). Every language is one file in screen_manager/translations, the same file the firmware
// and the add-on read (docs/TRANSLATING.md). The page takes its `editor` section, and for the mockup the words the screens
// themselves show (`screen.ha`, `time`, `date`, `number`, `tile.page`); the build leaves the rest out and compiles every
// text (vite.config.ts). English is the source and comes with the page. Another language loads the first time the
// editor or a mockup needs it, and a text it doesn't have yet shows in English.
import metas from "virtual:esp-screens-languages";
import { createI18n } from "vue-i18n";
import english from "../../screen_manager/translations/en.json";

export type LanguageMeta = { name: string; english: string; script?: string; plural: string; clock?: string; checked: boolean };
// A language's texts as the build hands them over: compiled, and only the page's part (not the file's own shape).
type Texts = Record<string, unknown>;
const en = english as unknown as Texts;

// Which form of a plural text ("1 hour ago | {n} hours ago") fits n, by the file's `_meta.plural`: the same rules as the
// screens (components/smart_display/screen_text_gen.py).
const few = (n: number) => n % 10 >= 2 && n % 10 <= 4 && (n % 100 < 12 || n % 100 > 14);
export const PLURAL_RULES: Record<string, (n: number) => number> = {
  one_other: (n) => (n === 1 ? 0 : 1),
  one_upto_1: (n) => (Math.abs(n) <= 1 ? 0 : 1),
  slavic_pl: (n) => (n === 1 ? 0 : few(n) ? 1 : 2),
  east_slavic: (n) => (n % 10 === 1 && n % 100 !== 11 ? 0 : few(n) ? 1 : 2),
  none: () => 0,
};

// Every other language by its code, as a chunk of its own; every language's _meta comes with the page.
const files = import.meta.glob<Texts>(
  ["../../screen_manager/translations/*.json", "!../../screen_manager/translations/en.json"], { import: "default" });
const LOADERS: Record<string, () => Promise<Texts>> = Object.fromEntries(
  Object.entries(files).map(([path, load]) => [path.slice(path.lastIndexOf("/") + 1, -".json".length), load]));
const META: Record<string, LanguageMeta> = { ...metas };
const LOADED = new Set(["en"]);
// A text with fewer forms than its language's rule uses its last one, as on the screens.
const RULES: Record<string, (choice: number, forms: number) => number> = {};
const addRule = (code: string) =>
  (RULES[code] = (choice, forms) => Math.min((PLURAL_RULES[META[code]?.plural] || PLURAL_RULES.one_other)(choice), forms - 1));

/** Every language the editor has, English first. */
export const languages = () => ["en", ...Object.keys(META).filter((code) => code !== "en").sort()];
languages().forEach(addRule);

export const i18n = createI18n({
  legacy: false,
  locale: "en",
  fallbackLocale: "en",
  messages: { en } as Record<string, any>,
  pluralRules: RULES,
  // A language without a text falls back to English quietly. Only a key English lacks is a mistake, told while developing.
  missingWarn: false,
  fallbackWarn: false,
  missing: (locale, key) => {
    if (import.meta.env.DEV && locale === "en") console.warn(`[i18n] ${key} is not in en.json`);
  },
});
export const t = i18n.global.t;
/** Whether a text exists, in English when no language is named. */
export const te = (key: string, locale = "en"): boolean => (i18n.global.te as (key: string, locale?: string) => boolean)(key, locale);
export const editorLanguage = () => i18n.global.locale.value;
/** A language's `_meta`: its own name, its name in English, its plural rule, whether it is checked. */
export const languageMeta = (code: string): LanguageMeta | undefined => META[code];

// Home Assistant opens the page in its ingress iframe, on its own origin, and sets its <html lang> to the language of
// the user's profile. Elsewhere (SCREEN_DEV, `npm run dev`) the browser's language, and English after that.
export function requestedLanguage(win: Window = window) {
  try {
    if (win.parent !== win) {
      const lang = win.parent.document.documentElement.lang;
      if (lang) return lang;
    }
  } catch {
    // A parent on another origin keeps its document to itself.
  }
  return win.navigator.language || "en";
}
// The language of the list that fits a code best: the same one (pt-BR), else its base language (pt), else none.
export function matchLanguage(wanted: string | null | undefined, available = languages()) {
  const find = (code: string) => available.find((own) => own.toLowerCase() === code.toLowerCase());
  const code = (wanted || "").trim();
  return code ? find(code) || find(code.split("-")[0]) : undefined;
}
export const pickLanguage = (wanted: string | null | undefined, available = languages()) => matchLanguage(wanted, available) || "en";

/** A language's texts, as the build hands them over; `meta` for a language without a file (the tests). */
export function addLanguage(code: string, texts: Texts, meta: LanguageMeta | undefined = META[code]) {
  if (meta) META[code] = meta;
  if (!RULES[code]) addRule(code);
  i18n.global.setLocaleMessage(code, texts as any);
  LOADED.add(code);
}
const loading: Record<string, Promise<boolean>> = {};
/** Loads a language once; true when the editor has it. */
export function loadLanguage(code: string): Promise<boolean> {
  if (LOADED.has(code)) return Promise.resolve(true);
  const load = LOADERS[code];
  if (!load) return Promise.resolve(false);
  return (loading[code] ??= load().then(
    (texts) => (addLanguage(code, texts), true),
    () => (delete loading[code], false),
  ));
}
/** The editor's own language: the one of the user's Home Assistant profile, as far as ESP Screens has it. */
export async function setEditorLanguage(code = pickLanguage(requestedLanguage())) {
  const chosen = (await loadLanguage(code)) ? code : "en";
  i18n.global.locale.value = chosen;
  document.documentElement.lang = chosen;
  return chosen;
}

// How the screens write a number (app 0.2.90), as the firmware does (screen_text::localize): the decimal mark, the mark
// between thousands, and from how many digits a whole number gets them. The choice under Language & region is one of
// three styles; "auto" writes numbers as the screens' language does (its screen.number).
export type NumberStyle = "point" | "comma" | "space";
export type NumberMarks = { decimal: string; group: string; from: number };
export const STYLE_MARKS: Record<NumberStyle, NumberMarks> = {
  point: { decimal: ".", group: ",", from: 4 }, comma: { decimal: ",", group: ".", from: 4 }, space: { decimal: ",", group: " ", from: 4 },
};
/** The marks of a language's own numbers: "1.234,5" in Dutch, 1234 but 12.345 in Italian (CLDR's minimum grouping). */
export function languageMarks(locale: string): NumberMarks {
  const say = (key: string) => t(`screen.number.${key}`, {}, { locale });
  return { decimal: say("decimal"), group: say("group"), from: Number(say("group_min")) >= 2 ? 5 : 4 };
}
/** A state as Home Assistant sends a number ("1234.5", "-3"), written with these marks; any other text as it is. */
export function numberText(value: string | number, marks: NumberMarks) {
  const text = String(value), plain = /^(-?)(\d+)(?:\.(\d+))?$/.exec(text);
  if (!plain) return text;
  const whole = plain[2].length >= marks.from ? plain[2].replace(/\B(?=(\d{3})+$)/g, marks.group) : plain[2];
  return `${plain[1]}${whole}${plain[3] !== undefined ? marks.decimal + plain[3] : ""}`;
}

// The time and number format of the user's own Home Assistant profile, where it names one: a 12- or 24-hour clock, and
// "1,234.5" (comma_decimal), "1.234,5" (decimal_comma) or "1 234,5" (space_comma). "Follow the language", "use the
// system's" and "none" name none. Only inside Home Assistant's page, whose <home-assistant> element holds them.
const PROFILE_NUMBERS: Record<string, NumberStyle> = { comma_decimal: "point", decimal_comma: "comma", space_comma: "space" };
export function haProfile(win: Window = window): { clock?: "12" | "24"; numbers?: NumberStyle } {
  try {
    if (win.parent === win) return {};
    const locale = (win.parent.document.querySelector("home-assistant") as any)?.hass?.locale;
    const clock = ["12", "24"].includes(locale?.time_format) ? locale.time_format : undefined;
    return { ...(clock ? { clock } : {}), ...(PROFILE_NUMBERS[locale?.number_format] ? { numbers: PROFILE_NUMBERS[locale.number_format] } : {}) };
  } catch {
    return {};
  }
}

// "1, 2 and 3", the way the language writes a list (the CLDR, from the browser). The English texts write it without a
// comma before "and", as British English does.
export const andList = (items: (string | number)[], locale = editorLanguage()) =>
  new Intl.ListFormat(locale === "en" ? "en-GB" : locale, { type: "conjunction" }).format(items.map(String));
